### PR TITLE
EU-100

### PR DESCRIPTION
**Pull request by anton-serozhechkin**

### CODE REVIEW

File: apps/educational_institutions/models.py [hunk 1]
```diff
@@ -33,10 +33,10 @@ class Faculty(Base):
     name = Column(VARCHAR(length=255), nullable=False)
     shortname = Column(VARCHAR(length=20))
     main_email = Column(VARCHAR(length=50))
-    dekan_id = Column(INTEGER, ForeignKey('dekan.dekan_id'))
+    dean_id = Column(INTEGER, ForeignKey('dean.dean_id'))
```
Comment (YehorVoytenko) [inline apps/educational_institutions/models.py:36]: rename dean to dekan

**Review by YehorVoytenko — DISMISSED** [review]
Approved)

File: apps/educational_institutions/models.py (hunk 1 above)
Comment (anton-serozhechkin) [reply apps/educational_institutions/models.py:36]: not needed

File: apps/common/utils.py [hunk 2]
```diff
@@ -36,7 +38,7 @@ def get_student_attr(student):
             message="A user account already exists. Please check your email for details.",
             code=http_status.HTTP_409_CONFLICT
         )
-    return student.full_name, student.faculty_id
+    return student.last_name, student.first_name, student.faculty_id
```
Comment (Kostiantyn-Salnykov) [inline apps/common/utils.py]: Should we use: First name and then Last name

File: apps/educational_institutions/models.py [hunk 3]
```diff
@@ -63,28 +63,34 @@ def __repr__(self):
...
-    full_name = Column(VARCHAR(length=255), nullable=False)
+    last_name = Column(VARCHAR(length=50), nullable=False)
+    first_name = Column(VARCHAR(length=50), nullable=False)
+    middle_name = Column(VARCHAR(length=50))
 
     university = relationship("University", back_populates="rector")
 
     def __str__(self):
-        return f'{self.__class__.__name__}(rector_id="{self.rector_id}",full_name="{self.full_name}")'
+        return f'{self.__class__.__name__}(rector_id="{self.rector_id}", last_name="{self.last_name}",' \
```
Comment (Kostiantyn-Salnykov) [inline apps/educational_institutions/models.py]: Should we use: First -> Middle -> Last names
E.G. Ivan Ivanovich Ivanchenko

File: apps/educational_institutions/models.py [hunk 4]
```diff
@@ -103,8 +109,9 @@ def __repr__(self):
                           Column('shortname', VARCHAR(20)),
                           Column('main_email', VARCHAR(50)),
                           Column('university_id', INTEGER),
-                          Column('dekan_id', INTEGER),
-                          Column('dekan_full_name', VARCHAR(255)))
+                          Column('dean_id', INTEGER),
+                          Column('dean_full_name', JSON))
```
Comment (Kostiantyn-Salnykov) [inline apps/educational_institutions/models.py:113]: Why did you add this field as JSON type?

File: apps/services/schemas.py [hunk 5]
```diff
@@ -24,17 +24,17 @@ class UserRequestExistenceOut(BaseOutSchema):
...
 
 class UserRequestBookingHostelOut(BaseOutSchema):
-    full_name: str
+    full_name: Dict[str, str]
     user_id: int
     faculty_name: str
     university_id: int
     short_university_name: str
-    rector_full_name: str
+    rector_full_name: Dict[str, str]
```
Comment (Kostiantyn-Salnykov) [inline apps/services/schemas.py]: Should be an embedded schema, that describes this structure

File: apps/services/schemas.py [hunk 6]
```diff
@@ -105,7 +105,7 @@ class HostelAccomodationViewOut(BaseOutSchema):
     university_name: str
     organisation_code: str
     payment_recognation: str  # TODO spelling error
-    commandant_full_name: str
+    commandant_full_name: Dict[str, str]
```
Comment (Kostiantyn-Salnykov) [inline apps/services/schemas.py]: Should be an embedded schema, that describes this structure

File: apps/hostel/schemas.py [hunk 7]
```diff
@@ -10,7 +12,7 @@ class HostelListOut(BaseOutSchema):
     street: str
     build: str
     commandant_id: int
-    commandant_full_name: str
+    commandant_full_name: Dict[str, str]
```
Comment (Kostiantyn-Salnykov) [inline apps/hostel/schemas.py]: Should be an embedded schema, that describes this structure

File: apps/educational_institutions/schemas.py [hunk 8]
```diff
@@ -45,7 +45,7 @@ class FacultyOut(BaseOutSchema):
     shortname: str
     main_email: str = None
     university_id: int
-    dekan_full_name: str = None
+    dean_full_name: Dict[str, str] = None
```
Comment (Kostiantyn-Salnykov) [inline apps/educational_institutions/schemas.py]: Should be an embedded schema, that describes this structure

**Review by Kostiantyn-Salnykov — CHANGES_REQUESTED** [review]

File: apps/educational_institutions/schemas.py (hunk 8 above)
Comment (anton-serozhechkin) [reply apps/educational_institutions/schemas.py]: example?

File: apps/hostel/schemas.py (hunk 7 above)
Comment (anton-serozhechkin) [reply apps/hostel/schemas.py]: example?

File: apps/services/schemas.py (hunk 6 above)
Comment (anton-serozhechkin) [reply apps/services/schemas.py]: example?

File: apps/services/schemas.py (hunk 5 above)
Comment (anton-serozhechkin) [reply apps/services/schemas.py]: example?

File: apps/educational_institutions/models.py (hunk 4 above)
Comment (anton-serozhechkin) [reply apps/educational_institutions/models.py:113]: see changes im main.sql file

File: apps/educational_institutions/schemas.py (hunk 8 above)
Comment (anton-serozhechkin) [reply apps/educational_institutions/schemas.py]: fixed

File: apps/hostel/schemas.py (hunk 7 above)
Comment (anton-serozhechkin) [reply apps/hostel/schemas.py]: fixed

File: apps/services/schemas.py (hunk 6 above)
Comment (anton-serozhechkin) [reply apps/services/schemas.py]: fixed

File: apps/services/schemas.py (hunk 5 above)
Comment (anton-serozhechkin) [reply apps/services/schemas.py]: fixed

File: apps/educational_institutions/models.py (hunk 3 above)
Comment (anton-serozhechkin) [reply apps/educational_institutions/models.py]: fixed

File: apps/common/utils.py (hunk 2 above)
Comment (anton-serozhechkin) [reply apps/common/utils.py]: fixed

File: apps/common/schemas.py [hunk 9]
```diff
@@ -40,3 +40,9 @@ class JSENDErrorOutSchema(JSENDOutSchema):
     status: JSENDStatus = Field(default=JSENDStatus.ERROR)
     data: Union[str, NoneType]
     code: int = Field(default=http_status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+
+class FullaNameSchema(BaseOutSchema):
```
Comment (Kostiantyn-Salnykov) [inline apps/common/schemas.py]: FullaNameSchema -> FullNameSchema

File: apps/common/utils.py [hunk 10]
```diff
@@ -16,13 +16,15 @@
...
+    if len(last_name) < 4:
+        last_name = last_name.join(first_name)
+    transliterated_last_name = translit(last_name)
+    return get_login(transliterated_last_name)
 
 
-def get_generated_username(full_name):
-    transliterated_full_name = translit(full_name)
-    return get_login(transliterated_full_name)
+def get_login(data: str) -> str:
```
Comment (Kostiantyn-Salnykov) [inline apps/common/utils.py]: Are you sure about this function name?
Maybe we can find another better understanding name for it.

File: apps/users/schemas.py [hunk 11]
```diff
@@ -139,30 +139,38 @@ class DeleteUserIn(BaseInSchema):
...
-class StudentCheckExistanceOut(BaseOutSchema):
+class StudentCheckExistenceOut(BaseOutSchema):
     student: int
     token: str
     expires: datetime
 
 
 class CreateStudentIn(BaseInSchema):
-    full_name: str
+    last_name: str
```
Comment (Kostiantyn-Salnykov) [inline apps/users/schemas.py]: Add validations for fields (according to DB schema)

File: apps/users/schemas.py [hunk 12]
```diff
@@ -139,30 +139,38 @@ class DeleteUserIn(BaseInSchema):
     user_id: int
 
 
-class StudentCheckExistanceIn(BaseInSchema):   # TODO spelling error
-    full_name: str
+class StudentCheckExistenceIn(BaseInSchema):
+    last_name: str
```
Comment (Kostiantyn-Salnykov) [inline apps/users/schemas.py]: Add validations for fields (according to DB schema)

File: apps/common/schemas.py [hunk 13]
```diff
@@ -40,3 +40,9 @@ class JSENDErrorOutSchema(JSENDOutSchema):
     status: JSENDStatus = Field(default=JSENDStatus.ERROR)
     data: Union[str, NoneType]
     code: int = Field(default=http_status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+
+class FullNameSchema(BaseOutSchema):
+    last_name: str
```
Comment (Kostiantyn-Salnykov) [inline apps/common/schemas.py:58]: Add validations for fields (according to DB schema)
P.S. You can also add default, max_length, description, title, example,examples.

**Review by Kostiantyn-Salnykov — CHANGES_REQUESTED** [review]

File: apps/common/schemas.py (hunk 9 above)
Comment (anton-serozhechkin) [reply apps/common/schemas.py]: was already fixed

File: apps/common/utils.py (hunk 10 above)
Comment (anton-serozhechkin) [reply apps/common/utils.py]: of course no, bcs i asked Victor to change it, but we didn't find any good name for this function. i'll change it, ok

File: apps/users/schemas.py (hunk 11 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: Ok

File: apps/users/schemas.py (hunk 12 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: ok

**Review by victor-nuzhniy — DISMISSED** [review]

File: apps/common/schemas.py (hunk 13 above)
Comment (anton-serozhechkin) [reply apps/common/schemas.py:58]: no need, bcs it's out schema

File: apps/common/utils.py (hunk 10 above)
Comment (anton-serozhechkin) [reply apps/common/utils.py]: @victor-nuzhniy please rename this method

File: apps/users/schemas.py [hunk 14]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
...
+        if not value.istitle():
+            raise ValueError("The first letter must be uppercase!")
+        return value
+
+    @validator('first_name')
+    def validate_first_name(cls, value):
+        if not value:
+            raise ValueError("The student's name is mandatory!")
+        if not value.istitle():
+            raise ValueError("The first letter must be uppercase!")
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: first letter of what?

File: apps/users/schemas.py [hunk 15]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
...
-    def validate_full_name(cls, value):
-        full_name = value.split()
-        if not full_name or len(full_name) < 2:
-            raise ValueError("The student's name and surname are mandatory!")
+    @validator('last_name')
+    def validate_last_name(cls, value):
+        if not value:
+            raise ValueError("The student's surname is mandatory!")
+        if not value.istitle():
+            raise ValueError("The first letter must be uppercase!")
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: first letter of what?

File: apps/users/schemas.py [hunk 16]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
...
+class StudentCheckExistenceOut(BaseOutSchema):
     student: int
     token: str
     expires: datetime
 
 
 class CreateStudentIn(BaseInSchema):
-    full_name: str
-    telephone_number: str
+    last_name: str = Field(default="Петренко", max_length=50)
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: don't use Ukrainian language please

File: apps/users/schemas.py [hunk 17]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
...
+    first_name: str = Field(default="Петро", max_length=50)
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: don't use Ukrainian language please

File: apps/users/schemas.py [hunk 18]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
...
+            raise ValueError("The first letter must be uppercase!")
+        return value
 
-class StudentCheckExistanceOut(BaseOutSchema):
+    @validator('first_name')
+    def validate_first_name(cls, value):
+        if not value:
+            raise ValueError("The student's name is mandatory!")
+        if not value.istitle():
+            raise ValueError("The first letter must be uppercase!")
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: first letter of what?

File: apps/users/schemas.py [hunk 19]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
...
+    last_name: str = Field(default="Петренко", max_length=50)
+    first_name: str = Field(default="Петро", max_length=50)
+    telephone_number: str = Field(default="380979889988", max_length=12, min_length=12)
 
+    @validator('last_name')
+    def validate_last_name(cls, value):
+        if not value:
+            raise ValueError("The student's surname is mandatory!")
+        if not value.istitle():
+            raise ValueError("The first letter must be uppercase!")
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: first letter of what?

File: apps/users/schemas.py [hunk 20]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
     user_id: int
 
 
-class StudentCheckExistanceIn(BaseInSchema):   # TODO spelling error
-    full_name: str
-    telephone_number: str
+class StudentCheckExistenceIn(BaseInSchema):
+    last_name: str = Field(default="Петренко", max_length=50)
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: don't use Ukrainian language please

File: apps/users/schemas.py [hunk 21]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
...
+    first_name: str = Field(default="Петро", max_length=50)
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: don't use Ukrainian language please

File: apps/users/schemas.py [hunk 22]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
...
     speciality_id: int
     gender: str
 
-    @validator('full_name')
-    def validate_full_name(cls, value):
-        full_name = value.split()
-        if not full_name or len(full_name) < 2:
-            raise ValueError("The student's name and surname are mandatory!")
+    @validator('last_name')
+    def validate_last_name(cls, value):
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: the same method we use in StudentCheckExistenceIn class, can we re-use it?

File: apps/users/schemas.py [hunk 23]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
...
+        if not value:
+            raise ValueError("The student's surname is mandatory!")
+        if not value.istitle():
+            raise ValueError("The first letter must be uppercase!")
+        return value
+
+    @validator('first_name')
+    def validate_first_name(cls, value):
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: the same method we use in StudentCheckExistenceIn class, can we re-use it?

File: apps/users/schemas.py [hunk 24]
```diff
@@ -139,38 +139,70 @@ class DeleteUserIn(BaseInSchema):
...
+        if not value:
+            raise ValueError("The student's name is mandatory!")
+        if not value.istitle():
+            raise ValueError("The first letter must be uppercase!")
         return value
 
     @validator('telephone_number')
     def validate_telephone_number(cls, value):
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py]: the same method we use in StudentCheckExistenceIn class, can we re-use it?

File: apps/users/schemas.py (hunk 15 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: ok

File: apps/users/schemas.py (hunk 14 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: ok

File: apps/users/schemas.py (hunk 16 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: ok

File: apps/users/schemas.py (hunk 20 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: ok

File: apps/users/schemas.py (hunk 21 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: ok

File: apps/users/schemas.py (hunk 22 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: ok

File: apps/users/schemas.py (hunk 23 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: ok

File: apps/users/schemas.py (hunk 24 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: ok

**Review by enerush — DISMISSED** [review]

File: apps/common/utils.py [hunk 25]
```diff
@@ -16,13 +16,15 @@
...
+    if len(last_name) < 4:
+        last_name = last_name.join(first_name)
+    transliterated_last_name = translit(last_name)
+    return get_generated_login(transliterated_last_name)
 
 
-def get_generated_username(full_name):
-    transliterated_full_name = translit(full_name)
-    return get_login(transliterated_full_name)
+def get_generated_login(data: str) -> str:
```
Comment (anton-serozhechkin) [inline apps/common/utils.py]: bad naming, because it's the same as "get_generated_username"...
so please rename it to, for eg "add_random_digits_and_cut_username"

File: apps/users/handlers.py [hunk 26]
```diff
@@ -1,9 +1,9 @@
 from apps.authorization.services import get_hashed_password
 from apps.common.exceptions import BackendException
-from apps.common.utils import (get_login, get_student_attr, get_token_data, get_generated_username,
+from apps.common.utils import (get_generated_login, get_student_attr, get_token_data, get_generated_username,
```
Comment (anton-serozhechkin) [inline apps/users/handlers.py]: don't forget to rename method there is

File: apps/users/handlers.py [hunk 27]
```diff
@@ -56,7 +56,7 @@ async def create_user(
         created_user = await user_service.create(
             session=session,
             data={
-                "login": get_login(user.email),
+                "login": get_generated_login(user.email),
```
Comment (anton-serozhechkin) [inline apps/users/handlers.py]: don't forget to rename method there is

File: apps/users/schemas.py (hunk 24 above)
Comment (anton-serozhechkin) [reply apps/users/schemas.py]: this method already implemented in StudentCheckExistenceIn, please delete it from CreateStudentIn bcs it's not need

File: apps/users/schemas.py [hunk 28]
```diff
@@ -139,38 +139,51 @@ class DeleteUserIn(BaseInSchema):
...
-
     @validator('telephone_number')
     def validate_telephone_number(cls, value):
-        if not value:
-            raise ValueError('The phone number cannot be empty!')
-        elif len(value) != 12:
-            raise ValueError('The phone number must contain 12 digits!')
+        if not value.isdigit():
+            raise ValueError('The phone number must consist of digits!')
         return value
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py:189]: can u also please add checking of middle_name uppercase, if this field isn't empty and was passed in the request?

File: apps/users/schemas.py (hunk 24 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py]: ok

File: apps/common/utils.py (hunk 25 above)
Comment (victor-nuzhniy) [reply apps/common/utils.py]: ok

File: apps/users/handlers.py (hunk 26 above)
Comment (victor-nuzhniy) [reply apps/users/handlers.py]: ok

File: apps/users/schemas.py (hunk 28 above)
Comment (victor-nuzhniy) [reply apps/users/schemas.py:189]: ok

**Review by victor-nuzhniy — APPROVED** [review]

File: apps/users/schemas.py [hunk 29]
```diff
@@ -139,39 +139,53 @@ class DeleteUserIn(BaseInSchema):
...
+class StudentCheckExistenceOut(BaseOutSchema):
     student: int
     token: str
     expires: datetime
 
 
-class CreateStudentIn(BaseInSchema):
-    full_name: str
-    telephone_number: str
+class CreateStudentIn(StudentCheckExistenceIn):
```
Comment (anton-serozhechkin) [inline apps/users/schemas.py:176]: after inheriting from class StudentCheckExistenceIn validation of field first, last names and telephone doesn't work

**Review by victor-nuzhniy — APPROVED** [review]
